WIP: Support global.json on arm64 as well

omajid opened the pull request:
I am trying to get ASP.NET Core to build on linux/arm64 machines (not cross compiled on x86_64).

On such machines, global.json makes arcade download the arm64 SDK, then the x64 and/or x86 runtimes. Now there are two copies of hostfxr in the local SDK:

```
$ find .dotnet/host/
.dotnet/host/
.dotnet/host/fxr
.dotnet/host/fxr/5.0.0-alpha1.19415.3
.dotnet/host/fxr/5.0.0-alpha1.19415.3/libhostfxr.so
.dotnet/host/fxr/5.0.0-alpha1.19470.7
.dotnet/host/fxr/5.0.0-alpha1.19470.7/libhostfxr.so
```

One is an x64 binary, the other is the arm64 binary:

```
$ file .dotnet/host/fxr/5.0.0-alpha1.19415.3/libhostfxr.so
.dotnet/host/fxr/5.0.0-alpha1.19415.3/libhostfxr.so: ELF 64-bit LSB shared object, ARM aarch64, version 1 (SYSV), dynamically linked, BuildID[sha1]=824b0c0f19a62296726be7bd12cef294ee3732af, stripped
$ file .dotnet/host/fxr/5.0.0-alpha1.19470.7/libhostfxr.so
.dotnet/host/fxr/5.0.0-alpha1.19470.7/libhostfxr.so: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, BuildID[sha1]=511b71993bb9762f46d69f377145d622208165a5, stripped
```

Unfortunately, the higher version is the x64 binary. That can't be loaded or used on arm64. The ASP.NET Core build quickly runs into an error:

```
/home/omajid/.nuget/packages/microsoft.net.compilers.toolset/3.4.0-beta1-19456-03/tasks/netcoreapp2.1/Microsoft.CSharp.Core.targets(59,5): error : Failed to load �#|, error: /home/omajid/AspNetCore/.dotnet/host/fxr/5.0.0-alpha1.19470.7/libhostfxr.so: cannot open shared object file: No such file or directory [/home/omajid/AspNetCore/eng/tools/RepoTasks/RepoTasks.csproj]                       
```

This change removes architecture-specific download versions from global.json. This lets [arcade download the arch-appropriate version of .NET Core](https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Arcade.Sdk/src/InstallDotNetCore.cs) which then lets the build continue past this error.

dougbu commented on global.json:
```diff
@@ -5,10 +5,7 @@
-        "$(MicrosoftNETCoreAppRuntimeVersion)"
-      ],
-      "dotnet/x86": [
+      "dotnet": [
```
```suggestion
      "dotnet/x64": [
        "$(MicrosoftNETCoreAppRuntimeVersion)"
      ],
      "dotnet/x86": [
        "$(MicrosoftNETCoreAppRuntimeVersion)"
      ],
      "dotnet/arm64": [
```

Parts of the full build need to cross-compile and test on alternative platforms. So, let's instead download it all.

omajid replied:
I was afraid of this. This is the first patch I tried. Unfortunately, that doesn't work :(

```
Setting msbuild verbosity to minimal                                                                                  
Downloading 'https://dot.net/v1/dotnet-install.sh'                                                                    
dotnet_install: Warning: Unable to locate libunwind. Probable prerequisite missing; install libunwind.                
dotnet-install: Downloading link: https://dotnetcli.azureedge.net/dotnet/Sdk/5.0.100-alpha1-013788/dotnet-sdk-5.0.100-alpha1-013788-linux-arm64.tar.gz
dotnet-install: Extracting zip from https://dotnetcli.azureedge.net/dotnet/Sdk/5.0.100-alpha1-013788/dotnet-sdk-5.0.100-alpha1-013788-linux-arm64.tar.gz
dotnet-install: Adding to current process PATH: `/home/omajid/AspNetCore/.dotnet`. Note: This change will be visible only when sourcing script.
dotnet-install: Installation finished successfully.
  Restore completed in 214.81 ms for /home/omajid/.nuget/packages/microsoft.dotnet.arcade.sdk/1.0.0-beta.19462.4/tools/Tools.proj.
dotnet_install: Warning: Unable to locate libunwind. Probable prerequisite missing; install libunwind.
dotnet-install: Downloading link: https://dotnetcli.azureedge.net/dotnet/Runtime/5.0.0-alpha1.19470.7/dotnet-runtime-5.0.0-alpha1.19470.7-linux-x64.tar.gz
dotnet-install: Extracting zip from https://dotnetcli.azureedge.net/dotnet/Runtime/5.0.0-alpha1.19470.7/dotnet-runtime-5.0.0-alpha1.19470.7-linux-x64.tar.gz
dotnet-install: Adding to current process PATH: `/home/omajid/AspNetCore/.dotnet`. Note: This change will be visible only when sourcing script.
dotnet-install: Installation finished successfully.
dotnet_install: Warning: Unable to locate libunwind. Probable prerequisite missing; install libunwind.
dotnet-install: .NET Core Runtime version 5.0.0-alpha1.19470.7 is already installed.
dotnet-install: Adding to current process PATH: `/home/omajid/AspNetCore/.dotnet`. Note: This change will be visible only when sourcing script.
dotnet-install: Installation finished successfully.
  Restore completed in 613.33 ms for /home/omajid/AspNetCore/eng/tools/RepoTasks/RepoTasks.csproj.
/home/omajid/.nuget/packages/microsoft.net.compilers.toolset/3.4.0-beta1-19456-03/tasks/netcoreapp2.1/Microsoft.CSharp.Core.targets(59,5): error : Failed to load �#4, error: /home/omajid/AspNetCore/.dotnet/host/fxr/5.0.0-alpha1.19470.7/libhostfxr.so: cannot open shared object file: No such file or directory [/home/omajid/AspNetCore/eng/tools/RepoTasks/
RepoTasks.csproj]
/home/omajid/.nuget/packages/microsoft.net.compilers.toolset/3.4.0-beta1-19456-03/tasks/netcoreapp2.1/Microsoft.CSharp.Core.targets(59,5): error : The library libhostfxr.so was found, but loading it from /home/omajid/AspNetCore/.dotnet/host/fxr/5.0.0-alpha1.19470.7/libhostfxr.so failed [/home/omajid/AspNetCore/eng/tools/RepoTasks/RepoTasks.csproj]
/home/omajid/.nuget/packages/microsoft.net.compilers.toolset/3.4.0-beta1-19456-03/tasks/netcoreapp2.1/Microsoft.CSharp.Core.targets(59,5): error :   - Installing .NET Core prerequisites might help resolve this problem. [/home/omajid/AspNetCore/eng/tools/RepoTasks/RepoTasks.csproj]
/home/omajid/.nuget/packages/microsoft.net.compilers.toolset/3.4.0-beta1-19456-03/tasks/netcoreapp2.1/Microsoft.CSharp.Core.targets(59,5): error :      https://go.microsoft.com/fwlink/?linkid=2063370 [/home/omajid/AspNetCore/eng/tools/RepoTasks/RepoTasks.csproj]
Build failed (exit code '1'). 
```

As I said in the original PR comments, the `arm64` and `x64` runtimes are slightly different (`5.0.0-alpha1.19415.3` vs `5.0.0-alpha1.19470.7`)... or overwritten? The `x64` hostfxr is newer and wins at the end, resulting in anon-executible binary.

dougbu replied:
:ok: this won't work until `dotnet-install.sh` is fixed. @omajid please file a dotnet/cli issue because that seems to be where the problem lies.

Alternative suggestion below.

dougbu reviewed (changes requested):
Thanks @omajid

Just curious: What is your requirement to build on ARM64?

omajid commented:
> Just curious: What is your requirement to build on ARM64?

This is part of the larger work that we are doing on [source-build](https://github.com/dotnet/source-build/) to enable arm64 (and arm32, thought that's not my current goal) builds: https://github.com/dotnet/source-build/issues/750. The goal is that Linux distributions like Fedora should be able to build and run all of .NET Core on arm64 (not cross-compiled).

dougbu commented:
@livarcocc and @dagood who owns the `dotnet-install` scripts and the naming of the runtime packs?

It seems like some runtimes (arm64 and x86) are not available for Linux and the scripts skip them. However an arm64 SDK is available, just not one containing the runtime we need.

omajid commented:
dotnet-install.sh is part of cli: https://github.com/dotnet/cli/blob/master/scripts/obtain/dotnet-install.sh

dagood commented:
> the naming of the runtime packs?

This is defined by https://github.com/dotnet/designs/pull/50, but doesn't seem relevant here as the Packs are only involved in SDK build and publish scenarios. The Runtime and SDK layouts are over at https://github.com/dotnet/designs/blob/master/accepted/install-locations.md.

> As I said in the original PR comments, the `arm64` and `x64` runtimes are slightly different (`5.0.0-alpha1.19415.3` vs `5.0.0-alpha1.19470.7`)... or overwritten? The `x64` hostfxr is newer and wins at the end, resulting in anon-executible binary.

I don't have any underlying understanding of what Arcade or the dotnet install script are doing with the global.json, but it seems likely that the lower-versioned one is the one that the `5.0.100-alpha1-013788` SDK includes, and the higher one is based on `MicrosoftNETCoreAppRuntimeVersion` which seems like it would be updated more often.

omajid commented:
@livarcocc @dagood @dougbu Any thoughts on what the next steps need to be? I am happy to do the legwork.

dagood commented:
My guess would be that Arcade and/or the dotnet-install script needs a new feature to help with this situation, or something's broken. The next thing I'd try to do is isolate what is being installed and why, (why does it install `x64` rather than `x86` or `arm64` with version `MicrosoftNETCoreAppRuntimeVersion` when all are specified?), then file a dotnet/arcade or dotnet/cli issue depending on what's going on.

omajid commented:
Here's what I see on my end (linux, arm64):

<details>
<summary>
x86 is not supported, because it's Linux. So it's completely skipped.
</summary>

Arcade prints:

```
         Skipping installing x86 runtimes because this is a non-Windows platform and .NET Core x86 is not currently supported on any non-Windows platform.
```

This is from https://github.com/dotnet/arcade/blob/64a027b8d8b3a8e8e214bb00da02148335e3bd92/src/Microsoft.DotNet.Arcade.Sdk/src/InstallDotNetCore.cs#L95.

I am not sure how x86 and x64 combo works on Windows.
</details>

<details>
<summary>
<code>dotnet-install.sh</code> follows the <code>global.json</code> order. It installs the runtimes in order of appearance in the json file. 
</summary>

The position of `dotnet/x64` vs `dotnet/arm64` matters. If I put `dotnet/arm64` first, it works on my end (but would probably break on x64)

This does **not** work:
```
"runtimes": {
  "dotnet/x64": [
    "$(MicrosoftNETCoreAppRuntimeVersion)"
  ],
  "dotnet/x86": [
    "$(MicrosoftNETCoreAppRuntimeVersion)"
  ],
  "dotnet/arm64": [
    "$(MicrosoftNETCoreAppRuntimeVersion)"
  ]
}
```
The calls to `dotnet-install.sh` look like this:

```
.dotnet/dotnet-install.sh --version 5.0.100-alpha1-014696 --install-dir /home/omajid/aspnetcore-global-json/.dotnet
.dotnet/dotnet-install.sh --version 5.0.0-alpha1.19506.5 --install-dir /home/omajid/aspnetcore-global-json/.dotnet --architecture x64 --runtime dotnet --skip-non-versioned-files
.dotnet/dotnet-install.sh --version 5.0.0-alpha1.19506.5 --install-dir /home/omajid/aspnetcore-global-json/.dotnet --architecture arm64 --runtime dotnet --skip-non-versioned-files
```

This version of `global.json` works:

```
"runtimes": {
  "dotnet/arm64": [
    "$(MicrosoftNETCoreAppRuntimeVersion)"
  ],
  "dotnet/x64": [
    "$(MicrosoftNETCoreAppRuntimeVersion)"
  ],
  "dotnet/x86": [
    "$(MicrosoftNETCoreAppRuntimeVersion)"
  ]
}
```

The `dotnet-install.sh` script is called like this:

```
.dotnet/dotnet-install.sh --version 5.0.100-alpha1-014696 --install-dir /home/omajid/aspnetcore-global-json/.dotnet
.dotnet/dotnet-install.sh --version 5.0.0-alpha1.19506.5 --install-dir /home/omajid/aspnetcore-global-json/.dotnet --architecture arm64 --runtime dotnet --skip-non-versioned-files
.dotnet/dotnet-install.sh --version 5.0.0-alpha1.19506.5 --install-dir /home/omajid/aspnetcore-global-json/.dotnet --architecture x64 --runtime dotnet --skip-non-versioned-files
.dotnet/dotnet-install.sh --version 5.0.0-alpha1.19506.5 --install-dir /home/omajid/aspnetcore-global-json/.dotnet --architecture arm64 --runtime dotnet --skip-non-versioned-files
.dotnet/dotnet-install.sh --version 5.0.0-alpha1.19506.5 --install-dir /home/omajid/aspnetcore-global-json/.dotnet --architecture x64 --runtime dotnet --skip-non-versioned-files
```

</details>

<details>
<summary>If the runtime - irrespective of the architecture - is already installed, `dotnet-install.sh` does nothing.</summary>

The invocations to `dotnet-install.sh` look like this:

```
.dotnet/dotnet-install.sh --version 5.0.100-alpha1-014696 --install-dir /home/omajid/aspnetcore-global-json/.dotnet
.dotnet/dotnet-install.sh --version 5.0.0-alpha1.19506.5 --install-dir /home/omajid/aspnetcore-global-json/.dotnet --architecture x64 --runtime dotnet --skip-non-versioned-files
.dotnet/dotnet-install.sh --version 5.0.0-alpha1.19506.5 --install-dir /home/omajid/aspnetcore-global-json/.dotnet --architecture arm64 --runtime dotnet --skip-non-versioned-files
```

And for the last (`arm64`) invocation, `dotnet-install` says:

```
dotnet-install: .NET Core Runtime version 5.0.0-alpha1.19506.5 is already installed
```
</details>

I guess my main take-away is that `dotnet-install.sh` is completely unprepared to handle multiple installations of different architectures in the same location.

Does anyone know how this works on Windows with `x86` and `x64`? Is it simply that `x64` gets installed and `x86` gets skipped?

Is this usage of `dotnet-install.sh` common? Should I try and modify `dotnet-install.sh` to handle multiple architectures? Or should I try and modify ASP.NET Core try and put multiple installations in dfferent locations?

dougbu commented:
> Does anyone know how this works on Windows with x86 and x64? Is it simply that x64 gets installed and x86 gets skipped?

On Windows, the x64 SDK gets placed in `.dotnet/` and the x86 runtime is added in `.dotnet/x86/`. If the requested x64 runtime is also newer than the default in the SDK, it's also placed in `.dotnet/`, adding new files and folders as needed e.g. adding `.dotnet\shared\Microsoft.NETCore.App\5.0.0-alpha1.19506.5` plus the default `.dotnet\shared\Microsoft.NETCore.App\5.0.0-alpha1.19467.5`. `.dotnet\x86\shared\Microsoft.NETCore.App` contains only the `5.0.0-alpha1.19506.5` folder. In other words, on Windows two copies of the same runtime are installed but in different directories.

I suspect we have logic in various places to set environment variables and so on related to the dotnet install locations that understands "native goes in `.dotnet/` and additional **x86** runtime may be found in `.dotnet/x86/` but minimal coverage for the arm64 case. Not sure whether the installer is doing the wrong thing for a second ARM64 runtime (@livarcocc @dsplaisted ?), you're hitting a problem in the repo code that handles different runtimes (@jkotalik I think you've done the most with handling the added x86 runtime; any suggestions?), or this is a problem in the Arcade SDK (@markwilkie, thoughts)?

My additional guess is that `dotnet-install.sh` or Arcade (probably Arcade) always places the x64 runtime in `.dotnet/` even if the installed SDK is for ARM64.

So, @markwilkie is it possible to update that logic to instead place the matching runtime in `.dotnet/`? Would end up with the same layout on regular Linux and Windows as today but Linux ARM64 would look like
```
.dotnet/ - ARM64 SDK and latest ARM64 runtime
.dotnet/x64/ - latest x64 runtime (not sure whether the build will use this but it doesn't hurt)
```
instead of the current
```
.dotnet/ - ARM64 SDK and latest x64 runtime
```

omajid commented:
> My additional guess is that `dotnet-install.sh` or Arcade (probably Arcade) always places the x64 runtime in .dotnet/ even if the installed SDK is for ARM64.

From what I can tell, arcade simply calls `dotnet-install.sh` (or `dotnet-install.ps1`) and lets it do its job. `dotnet-install.sh` has no special logic other than refusing to replace an already existing runtime with the same runtime; it ignores architectures completely. It looks like it simply installs x64 first (because it appears first in `global.json`) an then refuses to install other runtimes because they already exist (ignoring architecture).

`dotnet-install.ps1` actually contains some logic to look at the architecture and select between `.dotnet/` and `.dotnet/x86` based on the architecture. But it only handles x86 as a special case.

> .dotnet/ - ARM64 SDK and latest ARM64 runtime
> .dotnet/x64/ - latest x64 runtime (not sure whether the build will use this but it doesn't hurt)

Would it make sense to install the current architecture (whatever that is, arm64 in my case) to `.dotnet` and install other variants (x86, x64, even arm64 in case of building on x64) in `.dotnet/${arch}/`?

dougbu commented on global.json:
```diff
@@ -5,10 +5,7 @@
-      "dotnet/x86": [
+      "dotnet": [
         "$(MicrosoftNETCoreAppRuntimeVersion)"
       ]
```
```suggestion
      ],
      "dotnet/x86": [
        "$(MicrosoftNETCoreAppRuntimeVersion)"
      ]
```
If I'm understanding things correctly, `dotnet-install.sh` and `dotnet-install.ps1` will install the runtime for the current architecture. Then, `dotnet-install.sh` will ignore the x86 request while `dotnet-install.ps1` will install that runtime in the .dotnet/x86/ folder.

Basically, this limits your change to `"dotnet/x64"` ➡️ `"dotnet"`.

omajid replied:
Thanks, that's a **great** idea!

pranavkm commented:
@omajid does this PR - https://github.com/dotnet/arcade/pull/4132 - supersede this change?

omajid commented:
> @omajid does this PR - dotnet/arcade#4132 - supersede this change?

Not really; we will that fix and then something like what @dougbu suggested above to be able to start a build on arm64 at all.

omajid committed:
Support global.json on arm64 as well

arcade uses the runtime section of global.json to decide which
architecture + runtime combination needs to be installed.

With https://github.com/dotnet/arcade/pull/4132 arcade can install
foreign SDKs in separate locations correctly.

This change, suggested by @dougbu, makes arcade always install the
runtime for the local architecture (which means it should work on arm64
and x64) as well as the x86 architecture (skipped on Linux).

This gets us a working SDK/Runtime combo on arm64.

omajid commented:
@dougbu Does this change look okay to you?

pranavkm commented:
@omajid @dougbu's out this week. Might be a few more days until this gets seen.

dougbu commented:
@aspnet-hello triage https://dev.azure.com/dnceng/public/_build/results?buildId=399149

aspnet-hello commented:
This comment was made automatically. If there is a problem contact aspnetcore-build@microsoft.com.

I've triaged the above build. I've created/commented on the following issue(s)
https://github.com/aspnet/AspNetCore-Internal/issues/3327

dougbu reviewed (approved):
@omajid this looks good. Much appreciated!

I'll merge once the required builds pass.
